courses/additional_modules/labs/answers/advanced_privacy.txt

```diff
@@ -0,0 +1,157 @@
+
+package Messages is
+
+   type Message_T is private;
+
+   procedure Set_Content
+     (Message : in out Message_T;
+      Value   :        Integer);
+   function Content
+     (Message : Message_T)
+      return Integer;
+   function Image
+     (Message : Message_T)
+      return String;
+
+private
+   type Message_Content_T;
+   type Message_T is access Message_Content_T;
+
+end Messages;
+
+package body Messages is
+
+   type Message_Content_T is new Integer;
+
+   procedure Set_Content
+     (Message : in out Message_T;
+      Value   :        Integer) is
+      New_Value : constant Message_Content_T := Message_Content_T (Value);
+   begin
+      if Message = null
+      then
+         Message := new Message_Content_T'(New_Value);
+      else
+         Message.all := New_Value;
+      end if;
+   end Set_Content;
+
+   function Content
+     (Message : Message_T)
+      return Integer is (Integer (Message.all));
+
+   function Image
+     (Message : Message_T)
+      return String is ("**" & Message_Content_T'Image (Message.all));
+
+end Messages;
+
+package Messages.List_Types is
+
+   type List_T is private;
+
+private
+
+   type List_Content_T;
+   type List_T is access List_Content_T;
+
+   type Id_Type is range 1_000 .. 9_999;
+   type List_Content_T is record
+      Id      : Id_Type;
+      Content : Message_T;
+      Next    : List_T;
+   end record;
+
+end Messages.List_Types;
+
+with Messages.List_Types;
+package Messages.List_Types.Operations is
+
+   procedure Append
+     (List : in out List_T;
+      Item :        Message_T);
+   function Next
+     (List : List_T)
+      return List_T;
+   function Is_Null
+     (List : List_T)
+      return Boolean;
+   function Image
+     (Message : List_T)
+      return String;
+
+end Messages.List_Types.Operations;
+
+package body Messages.List_Types.Operations is
+
+   Id : Id_Type := Id_Type'First;
+
+   procedure Append
+     (List : in out List_T;
+      Item :        Message_T) is
+   begin
+      if List = null
+      then
+         List := new List_Content_T'(Id => Id, Content => Item, Next => null);
+      else
+         List.Next :=
+           new List_Content_T'(Id => Id, Content => Item, Next => null);
+      end if;
+      Id := Id_Type'Succ (Id);
```

Comment:
I haven't seen a check for `Id` elsewhere in this file. Should I assume it's acceptable for this implementation to raise an exception when the list is full? Otherwise, a check here (and probably a `Success` output parameter) would be nice.

Reply:
Not really worried about error handling in this lab - just focusing on the topic at hand.
To make these examples "good" code, it will obscure the subject matter (IMO)